README
    Author: mhr
    Posts: 37
    Joined: 07 May 2012, 10:12
    E-book readers owned: onyx-boox-m92 sony-trs-t1
    Number of books owned: 500
    Country: Germany

Dewarping with a regular calibration point grid

Post by mhr » 28 May 2012, 06:38
Hello all,

I refer to the post viewtopic.php?f=17&t=113&hilit=calibration, in which a calibration point grid is suggested to do a calibrated dewarping if the camera has a fixed position with respect to the scanned page. There are other approaches like viewtopic.php?f=13&t=784 to automatically recognize calibration information for dekeystoning. The latter seems appropriate for setups where the relative position is changing during scanning.

I have written a small command line program, which is capable to correct keystoning, page warping and camera lense distortion automatically by using a calibration point grid as suggested in the first referenced post from above. In the following posts I will discuss the approach taken and finally will supply the program, which is GPLed.

The general setup is the following:

Your scanning setup guarantees that the camera is in fixed geometrical position with respect to the platen. In contrast to all other setups, the platen need not be flat! A non-flat platen seems strange at first sight, but can be advantageous to avoid blur. I have estimated an absolute depth of focus of about 1 cm for a camera like mine (a Powershot A490), which seems to be roughly independent from the distance of the camera and the platen. Therefore, with the pythagorean theorem you can show, that noticeable blurring occurs quite fast for usual camera setups. If dR denotes the maximal depth of focus, R the vertical camera to platen distance and r the radius of a circle on a page with focused content, then the formula

Code: Select all

r = sqrt(dR * (2*R + dR))

gives the relationship between these values, providing that the camera is orthogonal to the platen. Some examples for dR = 1 cm:

    R = 44 cm ==> r = 9.4 cm
    R = 60 cm ==> r = 11 cm
    R = 80 cm ==> r = 12.6 cm
    R = 100 cm ==> r = 14.1 cm

If the platen is flat, 2*r is the maximal diameter of a page fully in focus, but if
the platen is part of a cylindrical surface, 2*r is the maximal height of a page fully in focus!
Therefore I plan to construct a scanner with a cylindrical platen to enhance sharpness.
The distortion will be no problem for me as the following posts should demonstrate.


CHANGELOG
Dewarping with a regular calibration point grid
https://forum.diybookscanner.org/viewtopic.php?f=19&t=2589
GNU GPL Version 3

Version 1.6 of the library libppmrose.a

    create library libppmrose.a

Version 1.5 of the programs ppmunwarp and ppmwhitebalance

    ppmwhitebalance was enhanced to accept a rectangle
    by using the new options -gx, -gy, -gw and -gh.

Version 1.4 together with the library

    splitted the identical parts of the programs ppmunwarp
    and ppmwhitebalancing into a library
    and eliminated a bug in ppmunwarp due to regression

Version 1.3 of ppmunwarp

    recognize and convert all valid raw ppm files

Version 1.2 of ppmunwarp

    just another update with a minor bugfix

The version 1.1 of ppmunwarp with PPI features

    is presented to obtain PPI information

Version 1.0 of ppmunwarp

    ppmunwarp, which does the actual calibration


src/ppmroselib.h
// ppmroselib.h
// Copyright (C) 2013 Michael Rose

// ============================================================================
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
// ============================================================================

// Include files; no other libraries needed than 'libgpp':

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdarg.h>
#include <math.h>
#include <new>

#define VERSION "1.5"

// ----------------------------------------------------------------------------

// Example to compile successfully under Visual C++ with
// cl -GX ppmroselib.cpp <prgname>.cpp

#ifdef _WIN32
  typedef unsigned __int8   u_int8_t;
  typedef unsigned __int16  u_int16_t;
  typedef          __int32    int32_t;
  typedef unsigned __int32  u_int32_t;
  inline double round(double z) { return floor(z+0.5); }
  #include <io.h>
  #include <fcntl.h>
  #define binmode(fh) _setmode(_fileno(fh),_O_BINARY)
#else
  #define binmode(fh)
#endif

// ----------------------------------------------------------------------------

// Global types:

typedef unsigned long ulong;
typedef u_int16_t     pixel_t;   // 16 bit RGB-values for pixel.
typedef u_int16_t     color_t;   // Value type for color       grid.
typedef int32_t       deform_t;  // Value type for deformation grid.

// ============================================================================
// General common functions.
// ============================================================================

// Error messages and normal messages (both on error channel):

void Error(const char *format, ...);
void Print(const char *format, ...);

// ============================================================================
// Unified handling of parameters/options.
// ============================================================================

// Options are defined via 'AddFlag', 'AddString', 'AddInt' and
// 'AddDouble' in the overloaded method 'Define()'. Additionally
// some concluding lines for the usage message can be supplied
// by 'ExtraUsage()'. The defining 'mode' has the following meaning:
//
// 0:  Initialize the option.
// 1:  Print option for debugging purpose.
// 2:  Calculate formatting tabulators for option list of usage method.
// 3:  Print option line for usage method.
// 4:  Read option from current command line argument.

class Parameter
{
  // Definition mode and suppression of normal messages:
  int mode, quiet;

  // Internal variables for option parsing:
  int  argi, argii, argc;
  char buf[256], *arg, **argv;

  void PrintOption();
  int  GetOption(const char *option, int extra);

protected:

  virtual void Define() { }
  virtual void Check()  { }

  int  AddFlag(const char *option, int& param,
               const char *show, const char *meaning);
  int  AddString(const char *option, const char *& param, const char *value,
                 const char *show, const char *meaning);
  int  AddInt(const char *option, int& param, int radix,
              int value, int low, int high,
              const char *show, const char *meaning);
  int  AddDouble(const char *option, double& param,
                 double value, double low, double high,
                 const char *show, const char *meaning);
  void ExtraUsage(const char *msg);

public:

  // Program name:
  const char *prgName;

  // Additional lonely input name:
  const char *inpName;

  Parameter(const char *prgName) { this->prgName = prgName; }

  void Debug();
  void Usage(int exitCode);
  void operator()(int argcPrg, char *argvPrg[]);
};

// ============================================================================
// Basic handling for raw PPM-images.
// ============================================================================

struct Image
{
  long     width,   // Image width.
           height,  // Image height.
           depth,   // Maximum color value per channel.
           size;    // 3 * width * height, each pixel has RGB-tripel.
  pixel_t *pixel;   // 3 * width * height RGB values (16 bit).

  Image()  { pixel = 0;  Reset(); }
  ~Image() {             Reset(); }

  void Reset(long width=0, long height=0, long depth=255);
  void Read(const char *filename, FILE *fh=0);
  void Write(const char *filename, FILE *fh, const char *creator);
};

// ============================================================================
// Constant/linear/cubic interpolation.
// ============================================================================

// Basic usage for interpolation 'value = f(x,y)' in 2D-array 'raw'
// with incremental offsets 'ox' and 'oy' in the corresponding dimensions
// ('c0' is a flag to force linear interpolation):
//
// Interpolation2 inter(c0, mx, ox, my, oy);
// <loop>
//   inter.cx.Set(x);  with 0.0 <= x <= mx
//   inter.cy.Set(y);  with 0.0 <= y <= my
//   value = inter(raw);

// 'f[4]' contains up to four interpolation coefficients, depending on
// the fraction between the aequidistant grid points. 'm' is the maximum
// possible grid index (between 0 and 'm'), 'o' the row/column offset
// to adjust the grid pointer for adjacent rows/columns and 'n' denotes
// the valid coefficients 'f[0 ... n]'.
// 'o2' is another offset suitable to create system matrices.
// The cubic interpolation is valid for 'm >= 2'.
// The interpolation changes automatically to linear order, if 'm < 2'
// and to constant order if 'm == 0'.

struct Interpolation1 {
  double f[4];
  long   order, m, mm, o, o2, i, n;
  void SetSize(int c0, long m) {
    this->m = m;
    order   = m < 2 ? m : (c0 ? 1 : 3);
    mm      = m ? m : 1; }
  long Offset() { return i * o; }
  Interpolation1(int c0, long m, long o) { SetSize(c0, m);  this->o = o; }
  Interpolation1(int c0, long m, long o, long o2) {
    SetSize(c0, m);  this->o = o;  this->o2 = o2; }
  void Set(double t1);
};

// ----------------------------------------------------------------------------

// Twodimensional constant/linear/cubic interpolation:

template<class ityp>
struct Interpolation2 {
  Interpolation1 cx,cy;  // Onedimensional interpolation along x- and y-axis.
  long           oo;
  long Offset() { return cx.Offset() + cy.Offset(); }
  Interpolation2(int c0, long mx, long ox, long my, long oy) :
    cx(c0, mx, ox), cy(c0, my, oy) { }
  Interpolation2(int c0, long mx, long ox, long my, long oy, long oo) :
    cx(c0, mx, ox, oo*ox), cy(c0, my, oy, oo*oy) { this->oo = oo; }
  double operator()(ityp *raw);
  void AssembleRhs(double scale, double *rhs);
  void AssembleMat(double *mat);
};

// ============================================================================
// Solver for symmetric positive definite systems with limited bandwidth.
// ============================================================================

// System modul to solve linear equation with band matrix.
// Matrix is of size 'n x n' with lower and upper bandwidth 'b'.
// The offset between valid entries on adjacent columns on the
// same row is 'o = 2*b', the adjacent diagonal elements have offset 'o + 1':

class BandedSystem
{
  long    n, b, o;
  double *sol, *rhs, *mat;

public:

  double *GetSol() { return sol; }
  double *GetRhs() { return rhs; }
  double *GetMat() { return mat; }

  BandedSystem()  { sol = 0;  Reset(); }
  ~BandedSystem() {           Reset(); }

  void Reset(long n=0, long b=0);
  void ClearRhs();
  void ClearMat();
  void Cholesky();
  void Solve();
};

// ============================================================================
// Main routine.
// ============================================================================

// A global 'Main'-routine is required:

extern void Main();
extern int  Handler(int argc, char *argv[], Parameter& param);


src/ppmroselib.cc
// ppmroselib.cc
// Copyright (C) 2013 Michael Rose

// ============================================================================
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
// ============================================================================

#include "ppmroselib.h"

// ============================================================================
// General common functions.
// ============================================================================

static int printFlag = 1;

// Error message with program termination:

#define ELEN 1024
void Error(const char *format, ...)
{
  static char buf[ELEN];  // Probably enough for error messages.
  buf[ELEN-5] = 'e';
  va_list ap;
  va_start(ap, format);
  vsnprintf(buf, ELEN, format, ap);
  va_end(ap);
  if (buf[ELEN-5])  strcpy(buf+(ELEN-5), " ...");
  throw buf;
}
#undef ELEN

// ----------------------------------------------------------------------------

// Normal messages, which can be suppressed:

void Print(const char *format, ...)
{
  if (printFlag) {
    va_list ap;
    va_start(ap, format);
    vfprintf(stderr, format, ap);
    va_end(ap);
} }

// ============================================================================
// Unified handling of global parameters/options.
// ============================================================================

// 'mode == 2' calculates column width of the formatted option list,
// 'mode == 3' prints the formatted option list:

void Parameter::PrintOption()
{
  char *ptr = buf;
  if (mode == 2) {
    if (argi < int(strlen(ptr)))  argi  = strlen(ptr);
    ptr += strlen(ptr) + 1;
    if (argii < int(strlen(ptr)))  argii = strlen(ptr);
    return; }
  // For 'mode == 3':
  fprintf(stderr, "%-*s ", argi,  ptr);  ptr += strlen(ptr) + 1;
  fprintf(stderr, "%-*s ", argii, ptr);  ptr += strlen(ptr) + 1;
  for (;;) {
    while (*ptr && *ptr != '\n')  fputc(*ptr++, stderr);
    if (!*ptr++)  break;
    fprintf(stderr, "\n%-*s %-*s ", argi, "", argii, ""); }
  fprintf(stderr, ".\n");
}

// ----------------------------------------------------------------------------

// Check if current option is equal to 'option' and get next argument
// if 'extra' is set. On success return 1:

int Parameter::GetOption(const char *option, int extra)
{
  if (argi != argii || strcmp(arg, option))  return 0;
  ++argi;
  if (extra) {
    if (argi >= argc)  Usage(1);
    arg = argv[argi++]; }
  return 1;
}

// ----------------------------------------------------------------------------

int Parameter::AddFlag(const char *option, int& param,
                       const char *show, const char *meaning)
{
  switch (mode) {
    case 0:   param = 0;                          break;
    case 1:   Print("%s = %d\n", option, param);  break;
    case 2:
    case 3:   sprintf(buf, "%s%c%s%c%s",
                      option, 0, (show?show:""), 0, meaning);
              PrintOption();  break;
    default:  if (GetOption(option, 0))  return (param = 1); }
  return 0;
}

// ----------------------------------------------------------------------------

int Parameter::AddString(const char *option, const char *& param,
                         const char *value,
                         const char *show, const char *meaning)
{
  switch (mode) {
    case 0:   param = value;                          break;
    case 1:   Print("%s = \"%s\"\n", option, param);  break;
    case 2:
    case 3:   sprintf(buf, "%s%c%s%c%s", 
                      option, 0, (show?show:value), 0, meaning);
              PrintOption();  break;
    default:  if (GetOption(option, 1)) { param = arg;  return 1; } }
  return 0;
}

// ----------------------------------------------------------------------------

int Parameter::AddInt(const char *option, int& param, int radix,
                      int value, int low, int high,
                      const char *show, const char *meaning)
{
  char *end;
  switch (mode) {
    case 0:   param = value;                      break;
    case 1:   Print("%s = %d\n", option, param);  break;
    case 2:
    case 3:   if (show)
                sprintf(buf, "%s%c%s%c%s", option, 0, show, 0, meaning);
              else if (radix == 16)
                sprintf(buf, "%s%c(%x)%c%s", option, 0, value, 0, meaning);
              else
                sprintf(buf, "%s%c(%d)%c%s", option, 0, value, 0, meaning);
              PrintOption();  break;
    default:  if (GetOption(option, 1)) {
                param = strtol(arg, &end, radix);
                if (*end || param < low || param > high)  Usage(1);
                return 1; } }
  return 0;
}

// ----------------------------------------------------------------------------

int Parameter::AddDouble(const char *option, double& param,
                         double value, double low, double high,
                         const char *show, const char *meaning)
{
  char *end;
  switch (mode) {
    case 0:   param = value;                          break;
    case 1:   Print("%s = %.16lg\n", option, param);  break;
    case 2:
    case 3:   if (show)
                sprintf(buf, "%s%c%s%c%s", option, 0, show, 0, meaning);
              else
                sprintf(buf, "%s%c(%.4lg)%c%s", option, 0, value, 0, meaning);
              PrintOption();  break;
    default:  if (GetOption(option, 1)) {
                param = strtod(arg, &end);
                // Accept rounding errors at boundary:
                if (param < low  && param >= low  - 1e-14)  param = low;
                if (param > high && param <= high + 1e-14)  param = high;
                if (*end || param < low || param > high)  Usage(1);
                return 1; } }
  return 0;
}

// ----------------------------------------------------------------------------

void Parameter::ExtraUsage(const char *msg)
{
  if (mode == 3)  fprintf(stderr, "\n%s\n", msg);
}

// ----------------------------------------------------------------------------

// Displays all options for debugging purpose:

void Parameter::Debug()
{
  Print("BEG Debugging output of options:\n");
  Print("Program name = \"%s\"\n", prgName);
  Print("Quiet flag = %d\n", quiet);
  Print("Input name = \"%s\"\n", inpName);
  mode = 1;
  Define();
  Print("END Debugging output of options.\n");
}

// ----------------------------------------------------------------------------

// Usage/help information:

void Parameter::Usage(int exitCode)
{
  if (exitCode < 0) {
    fprintf(stdout,
            "%s " VERSION "\n\n"
            "Copyright (C) 2013 Michael Rose\n"
            "License GPLv3+: GNU GPL version 3 or later"
            " <http://gnu.org/licenses/gpl.html>\n"
            "This is free software: you are free to change"
            " and redistribute it.\n"
            "There is NO WARRANTY, to the extent permitted by law.\n\n",
            prgName);
    exit(0); }
  fprintf(stderr,
          "Usage: %s [options] [--] [inpname or stdin]\n\nOptions:\n",
          prgName);
  for (mode = 2, argi=argii=0; mode < 4; ++mode) {
    sprintf(buf, "--version%c%cPrint program version", 0, 0);
    PrintOption();
    sprintf(buf, "-h%c%cPrint program usage", 0, 0);
    PrintOption();
    sprintf(buf, "-q%c%cSuppress normal program messages", 0, 0);
    PrintOption();
    Define(); }
  exit(exitCode);
}

// ----------------------------------------------------------------------------

// Sets global parameters/options from the command line:

void Parameter::operator()(int argcPrg, char *argvPrg[])
{
  int doOpt = 1;
  inpName = 0;
  argi    = 1;
  argc    = argcPrg;
  argv    = argvPrg;
  mode    = quiet = 0;
  // Set default values:
  Define();
  while (argi < argc) {
    arg = argv[argi];
    if (!*arg)  Usage(1);
    if (doOpt && *arg == '-') {
      if      (!strcmp(arg, "--"))         { doOpt = 0;  ++argi; }
      else if (!strcmp(arg, "-q"))         { quiet = 1;  ++argi; }
      else if (!strcmp(arg, "--version"))  Usage(-1);
      else if (!strcmp(arg, "-h"))         Usage(0);
      else {
        mode  = 4;
        argii = argi;
        Define();
        if (argi == argii)  Usage(1); } }
    else {
      if (++argi != argc || !*arg)  Usage(1);
      inpName = arg; } }
  printFlag = !quiet;
  // Additional checks and settings:
  Check();
}

// ============================================================================
// Basic handling for raw PPM-images.
// ============================================================================

// Resizes the image array by given size parameters:

void Image::Reset(long width, long height, long depth)
{
  if (pixel)  delete pixel;
  this->width  = width;
  this->height = height;
  this->depth  = depth;
  size         = 3 * width * height;
  pixel        = 0;
  if (size)  pixel = new pixel_t[size];
}

// ----------------------------------------------------------------------------

// Read raw PPM image file:

void Image::Read(const char *filename, FILE *fh)
{
  int      mode = 0, ch = 0;
  long     z = 0, n, m;
  u_int8_t buf[4096], *src;
  pixel_t  val,       *dst;
  if (fh != stdin)  fh = fopen(filename, "rb");
  else              binmode(fh);  // Windows.
  if (!fh)  Error("Couldn't read '%s' as raw PPM image file", filename);
  try {
    // Read preamble of PPM file:
    while (mode != 5) {
      // Do we need a next character?
      if      (mode & 64)                mode &= 63;
      else if ((ch = fgetc(fh)) == EOF)  throw 1;
      // Do we need to read space with comments?
      if (mode & 8) {
        if (isspace(ch)) {
          mode |= 16;  // We have seen at least one space/comment.
          if (ch == '\n')  mode &= 31;
          continue; }
        else if (ch == '#') { mode |= 48;  continue; }
        else {
          if (mode & 32)     continue;  // Ignore comment characters in line.
          if (!(mode & 16))  throw 1;   // No separation found?
          // Switch to number reading:
          mode &= 23; } }
      // Do we need to read a number?
      if (mode & 16) {
        if (!(mode & 32))  z = 0;  // Initialize at beginning.
        if (ch >= '0' && ch <= '9') {
          mode |= 32;  // We have seen at least one digit.
          z     = 10 * z + (ch - '0');
          if (z > 65535)  throw 1;
          continue; }
        if (!(mode & 32))  throw 1;  // We need at least one digit.
        // Switch to global parsing mode:
        mode &= 7; }
      // Main parsing of preamble:
      switch (mode) {
        case 0:   if (ch == 'P') { mode =  1;  continue; }  throw 1;
        case 1:   if (ch == '6') { mode = 10;  continue; }  throw 1;
        case 2:   width  = z;  mode = 75;  continue;
        case 3:   height = z;  mode = 76;  continue;
        default:  depth  = z;  mode =  5; } }
    // Now the preamble has been consumed, check for final space character:
    if (!isspace(ch) || !width || !height || !depth)  throw 1;
    if (width > 25000 || height > 25000 || width*height > 100000000)  throw 4;
    Reset(width, height, depth);
    // Read binary data; either 8bit or 16bit RGB channel values:
    n   = size;
    dst = pixel;
    if (depth > 255) {
      while (n) {
        n -= m = n < 2048 ? n : 2048;
        if (fread((src = buf), sizeof(u_int8_t) << 1, m, fh) != ulong(m))
          throw 2;
        while (m--) {
          val    = pixel_t(*src++);
          *dst++ = val = (val << 8) | pixel_t(*src++);
          if (val > depth)  throw 5; } } }
    else {
      while (n) {
        n -= m = n < 4096 ? n : 4096;
        if (fread((src = buf), sizeof(u_int8_t), m, fh) != ulong(m))
          throw 2;
        while (m--) {
          *dst++ = val = pixel_t(*src++);
          if (val > depth)  throw 5; } } }
    if (fgetc(fh) != EOF)  throw 3;
    if (fh != stdin)  fclose(fh); }
  catch (int nmr) {
    const char *msg;
    if (fh != stdin)  fclose(fh);
    switch (nmr) {
      case 1:   msg = "has wrong preamble";           break;
      case 2:   msg = "is too small";                 break;
      case 3:   msg = "is too big";                   break;
      case 4:   msg = "has too large dimensions";     break;
      default:  msg = "has inconsistent color depth"; }
    Error("PPM image file '%s' %s", filename, msg); }
}

// ----------------------------------------------------------------------------

// Write raw PPM image file:

void Image::Write(const char *filename, FILE *fh, const char *creator)
{
  long     n, m, k;
  u_int8_t buf[4096], *dst;
  pixel_t  val,       *src;
  if (fh != stdout)  fh = fopen(filename, "wb");
  else               binmode(fh);  // Windows.
  if (!fh)  Error("Couldn't write '%s' as raw PPM image file", filename);
  fprintf(fh, "P6\n");
  if (creator)  fprintf(fh, "# CREATOR: %s\n", creator);
  fprintf(fh, "%ld %ld\n%ld\n", width, height, depth);
  n   = size;
  src = pixel;
  if (depth > 255) {
    while (n) {
      n   -= m = k = n < 2048 ? n : 2048;
      dst  = buf;
      while (k--) {
        val    = *src++;
        *dst++ = u_int8_t(val >> 8);
        *dst++ = u_int8_t(val & 255); }
      fwrite(buf, sizeof(u_int8_t) << 1, m, fh); } }
  else {
    while (n) {
      n   -= m = k = n < 4096 ? n : 4096;
      dst  = buf;
      while (k--)  *dst++ = u_int8_t(*src++);
      fwrite(buf, sizeof(u_int8_t), m, fh); } }
  if (fh != stdout)  fclose(fh);
}

// ============================================================================
// Constant/linear/cubic interpolation.
// ============================================================================

// Initializes interpolation at location '0.0 <= t1 <= m'.
// The interpolation is C1-continuous if 'order == 3':

void Interpolation1::Set(double t1) {
  double t0 = floor(t1);
  long   r;
  i   = long(t0);
  t1 -= t0;
  if (i >= mm) { --i;  t1 += 1.0; }  // Border case.
  switch (order) {
    case 0:   f[n = r = 0] = 1.0;                   break;
    case 1:   f[r = 0] = 1.0 - t1;  f[n = 1] = t1;  break;
    default:
      t0 = 1.0 - t1;
      if (i) {
        if (i == m - 1) {  // Highest grid interval:
          f[    0] = -0.5 * t0 * t1;
          f[r = 1] =       t0 * (2.0 - t0);
          f[n = 2] = 0.5 * t1 * (2.0 - t0); }
        else {  // General grid interval:
          f[    0] = -0.5 * t0 * t0 * t1;
          f[r = 1] = -t0 * (((1.5 * t1) - 1.0) * t1 - 1.0);
          f[    2] = -t1 * (((1.5 * t0) - 1.0) * t0 - 1.0);
          f[n = 3] = -0.5 * t1 * t1 * t0; } }
      else {  // Lowest grid interval:
        f[r = 0] = 0.5 * t0 * (2.0 - t1);
        f[    1] =       t1 * (2.0 - t1);
        f[n = 2] = -0.5 * t0 * t1; } }
  i -= r;
}

// ----------------------------------------------------------------------------

// Returns interpolated value for previously specified grid point
// on array 'raw':
//
// The interpolation on a rectangular grid has
// the tensor product property.

template<class ityp>
double Interpolation2<ityp>::operator()(ityp *raw) {
  double  value = 0.0;
  long    ix,iy;
  ityp   *p;
  raw += Offset();
  for (ix=0; ix <= cx.n; ++ix, raw += cx.o) {
    for (iy=0, p=raw; iy <= cy.n; ++iy, p += cy.o) {
      value += double(*p) * (cx.f[ix] * cy.f[iy]); } }
  return value;
}

// ----------------------------------------------------------------------------

// Assembles right hand side vector "G' * rhs":

template <class ityp>
void Interpolation2<ityp>::AssembleRhs(double scale, double *rhs) {
  long   ix, iy;
  double *p, v;
  rhs += Offset();
  for (ix=0; ix <= cx.n; ++ix, rhs += cx.o) {
    v = scale * cx.f[ix];
    for (iy=0, p=rhs; iy <= cy.n; ++iy, p += cy.o) {
      *p += v * cy.f[iy]; } }
}

// ----------------------------------------------------------------------------

// Assembles full or band system matrix "G' * G":

template <class ityp>
void Interpolation2<ityp>::AssembleMat(double *mat) {
  long   ix, iy, ix2, iy2;
  double *p1, *p2, *p3, v1, v2, v3;
  mat += Offset() * (oo + 1);
  for (ix=0; ix <= cx.n; ++ix, mat += cx.o) {
    v1 = cx.f[ix];
    for (iy=0, p1=mat; iy <= cy.n; ++iy, p1 += cy.o) {
      v2   = v1 * cy.f[iy];
      for (ix2=0, p2=p1; ix2 <= cx.n; ++ix2, p2 += cx.o2) {
        v3 = v2 * cx.f[ix2];
        for (iy2=0, p3=p2; iy2 <= cy.n; ++iy2, p3 += cy.o2) {
          *p3 += v3 * cy.f[iy2]; } } } }
}

// ============================================================================
// Solver for symmetric positive definite systems with limited bandwidth.
// ============================================================================

void BandedSystem::Reset(long n, long b) {
  long    size;
  double *arr;
  if (sol)  delete sol;
  this->n = n;
  this->b = b;
  o       = b << 1;
  size    = (o + 3) * n;
  sol     = rhs = mat = 0;
  if (size) {
    sol = arr = new double[size];
    rhs = sol + n;
    mat = rhs + n;
    while (size--)  *arr++ = 0.0; }
}

// ----------------------------------------------------------------------------

void BandedSystem::ClearRhs() {
  long    s = n;
  double *p = rhs;
  while (s--)  *p++ = 0.0; }

// ----------------------------------------------------------------------------

void BandedSystem::ClearMat() {
  long    s = (o + 1) * n;
  double *p = mat;
  while (s--)  *p++ = 0.0; }

// ----------------------------------------------------------------------------

// Cholesky "C' * C = M" decomposition of band matrix (band is preserved):

void BandedSystem::Cholesky() {
  double *pi, *pj, *pji, *pkj, *pki, s;
  long    i, j, k, m, mo;
  for (i=0, mo=-o, pi=mat; i < n; ++i, pi += o + 1) {
    if (i <= b) { m = i;  mo += o; } else m = b;
    for (j = 0, pji=pi-m, pj=pji-mo; j <= m; ++j, ++pji, pj += o+1) {
      for (k=0, pkj=pj, pki=pji, s=*pji; k < j; ++k)  s -= *--pkj * *--pki;
      if (j == m) {
        if (s < 1e-14)  Error("System matrix not positive definite");
        *pji = sqrt(s); }
      else  *pji = s / *pj; } }
}

// ----------------------------------------------------------------------------

// Solve "C' * C * sol = rhs":

void BandedSystem::Solve() {
  double *ps, *pi, *pji, *pjs, s;
  long    i, j, m;
  for (i=0; i < n; ++i) sol[i] = rhs[i];
  // Solve for transposed cholesky factor:
  for (i=0, pi=mat, ps=sol; i < n; ++i, pi += o + 1) {
    m = i <= b ? i : b;
    for (j=0, pji=pi, pjs=ps, s=*ps; j < m; ++j)  s -= *--pji * *--pjs;
    *ps++ = s / *pi; }
  // Solve for cholesky factor:
  for (i=0, pi=mat+(n-1)*(o+1), ps=sol+(n-1); i < n; ++i, pi -= o + 1) {
    m = i <= b ? i : b;
    for (j=0, pji=pi+o, pjs=ps, s=*ps; j < m; ++j, pji += o)
      s -= *pji * *++pjs;
    *ps-- = s / *pi; }
}

// ============================================================================
// Main routine.
// ============================================================================

// Main program to supply command line args, catch errors and
// call the 'Main' routine:

int Handler(int argc, char *argv[], Parameter& param)
{
  try {
    try {
      param(argc, argv);
      Main(); }
    // Catch exception from 'new' operator:
    catch (std::bad_alloc& ba) { throw "Out of memory"; } }
  catch (const char *msg) {
    fprintf(stderr,"!!! Error in %s:\n!!! %s!\n", param.prgName, msg);
    return 1; }
  return 0;
}

// ============================================================================
// Explicit template instantiation.
// ============================================================================

template class Interpolation2<double>;
template class Interpolation2<pixel_t>;
template class Interpolation2<deform_t>;
